feat(ARRR): dockerize zombie/pirate tests

borngraced opened the pull request:
need review here https://github.com/KomodoPlatform/ZOMBIE_test_docker cc @dimxy

shamardy commented:
@borngraced what is left to make this ready for review. @onur-ozkan @dimxy please ignore reviewing this until it's ready.

borngraced commented:
> @borngraced what is left to make this ready for review. @onur-ozkan @dimxy please ignore reviewing this until it's ready.

Some minor adjustments to the zombie docker image still ongoing

dimxy commented on mm2src/mm2_main/tests/docker_tests/z_coin_docker_tests.rs:
```diff
@@ -0,0 +1,242 @@
+
+#[tokio::test(flavor = "multi_thread")]
+async fn zombie_coin_send_and_refund_maker_payment() {
+    let _lock = TEST_MUTEX.lock().await;
```
I think we should set this lock on the API level (not in tests only)
(to prevent [this issue with parallel calls](https://github.com/KomodoPlatform/librustzcash/blob/4e030a0f44cc17f100bf5f019563be25c5b8755f/zcash_client_backend/src/data_api/wallet.rs#L72-L73), right?

borngraced replied:
no sure how we can approach this btw.. since we need to wait from start to finish of a pirate swap.

dimxy replied:
I think we just should prevent a note to be selected the second time from the received_notes table.
It should not be selected after we update its `spent` status with the store_sent_tx fn.
So I guess, we need to set a lock before where we add notes and release it after the store_sent_tx fn called. The places in code where we add notes are: init_withdraw, z_send_dex_fee, z_send_htlc

dimxy replied:
I made a [sample code](https://github.com/KomodoPlatform/komodo-defi-framework/compare/fix-zombie-tests...dimxy:komodo-defi-framework:fix-zombie-tests?expand=1#) with Z_NOTE_LOCK in the send_outputs fn (it's not ready for prod, just to prove the idea of a lock to prevent access to the available notes)

I also had to add a simple wait_for_z_balance fn, otherwise I had insufficient balance error. So probably the  spendable_notes_required_for_tx fn does work properly when different ZCoin {} created in concurrent zombie tests or its confirmation logic is not sufficient: I noticed that transactions may stay for few blocks in mempool so I wait for 5 blocks in wait_for_z_balance.

borngraced commented on mm2src/mm2_test_helpers/src/for_tests.rs:
```diff
@@ -480,9 +477,9 @@ pub fn zombie_conf() -> Json {
+        "overwintered": 1,
         "mm2":1,
-        "avg_blocktime": 60,
+        "avg_blocktime": 20,
```
I wanted fast block production via docker image. That made think it might have side effect with this which is why I updated it to the same average blocktime from the docker image https://github.com/KomodoPlatform/ZOMBIE_test_docker/blob/6a06dcf1cf08c52f95a18466b5adf33ad387396e/start_blockchain.py#L30(should be 10 tho.)

I can revert if you think  the changes in the script above doesn't really matter.

dimxy replied:
Ah I see you changed ac_blocktime to 10.
avg_blocktime is used to estimate number of blocks from a duration, so I guess it's better to have it equal to ac_blocktime.

borngraced commented on mm2src/mm2_main/tests/docker_tests/z_coin_docker_tests.rs:
```diff
@@ -0,0 +1,242 @@
+    drop(_lock);
+}
+
+// TODO: fix test
+#[ignore]
```
Yeah, I'm absolutely going to attend to test in this PR

borngraced commented on mm2src/mm2_main/tests/docker_tests/docker_tests_common.rs:
```diff
@@ -115,6 +118,8 @@ pub static GETH_RPC_URL: &str = "http://127.0.0.1:8545";
 #[cfg(any(feature = "sepolia-maker-swap-v2-tests", feature = "sepolia-taker-swap-v2-tests"))]
 pub static SEPOLIA_RPC_URL: &str = "https://ethereum-sepolia-rpc.publicnode.com";
 
+pub const ZOMBIE_ASSET_DOCKER_IMAGE: &str = "docker.io/borngraced/zombietestrunner";
+pub const ZOMBIE_ASSET_DOCKER_IMAGE_WITH_TAG: &str = "docker.io/borngraced/zombietestrunner:multiarch";
```
the docker image needs to be reviewed first https://github.com/KomodoPlatform/ZOMBIE_test_docker

dimxy commented on mm2src/coins/z_coin.rs:
```diff
@@ -1866,7 +1934,7 @@ impl InitWithdrawCoin for ZCoin {
         };
 
-        let (tx, data, _sync_guard) = self.gen_tx(vec![], vec![z_output]).await?;
+        let (tx, _, data, _sync_guard) = self.gen_tx(vec![], vec![z_output]).await?;
```
Why is there no store_sent_tx() call for withdraw?

borngraced commented on mm2src/coins/z_coin/z_htlc.rs:
```diff
@@ -16,13 +15,21 @@ use crate::NumConversError;
+use zcash_client_backend::address::RecipientAddress;
+use zcash_client_backend::data_api::SentTransaction;
+use zcash_client_backend::wallet::AccountId;
+#[cfg(not(target_arch = "wasm32"))]
+use zcash_client_sqlite::error::SqliteClientError;
```
done

borngraced commented on mm2src/coins/z_coin.rs:
```diff
@@ -1083,6 +1120,37 @@ async fn z_coin_from_conf_and_params_with_z_key(
     builder.build().await
 }
 
+/// Initialize `ZCoin` with a forced `z_spending_key`.
+#[allow(clippy::too_many_arguments)]
+pub async fn z_coin_from_conf_and_params_with_docker(
```
done

borngraced commented:
this PR(https://github.com/KomodoPlatform/komodo-defi-framework/pull/2374) is slowly diverting from it's main purpose. Hence, I have removed additional fixes that shouldn't have been done here...removed fixes will
be re-implemented in `improve shielded transactions change notes handling in swaps PR.

dimxy commented:
Do I understand it right that we are going to solve the problems with concurrent zombie tests failure?
If so I am okay with this PR

dimxy reviewed (approved):
LGTM

onur-ozkan commented on mm2src/coins/z_coin.rs:
```diff
@@ -282,10 +280,14 @@ impl ZCoin {
-        )
+        let mut watcher = self.z_fields.sync_state_connector.lock().await;
+        while let Some(sync) = watcher.sync_watcher.next().await {
+            if matches!(sync, SyncStatus::Finished { block_number: _, .. }) {
```
```suggestion
            if matches!(sync, SyncStatus::Finished { .. }) {
```

onur-ozkan reviewed (approved):
LGTM with some minors

shamardy reviewed (commented):
First iteration!

borngraced commented on mm2src/mm2_test_helpers/src/for_tests.rs:
```diff
@@ -478,9 +478,9 @@ pub fn zombie_conf() -> Json {
+        "overwintered": 1,
         "mm2":1,
-        "avg_blocktime": 60,
+        "avg_blocktime": 10,
```
We are allowed to  tweak this in komodod for test so I did for faster block generations.

shamardy replied:
As discussed, it's used in non dockarized tests as well such as https://github.com/KomodoPlatform/komodo-defi-framework/blob/927f84b4e5db6d0a0c8c4cc4b2e5aefdb7a39599/mm2src/mm2_main/tests/mm2_tests/z_coin_tests.rs#L163-L167
I suggest having different config for each with shared configs.

borngraced replied:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2374/commits/642f493f6426881f529fa9fbb95dc45ef9c0af15

borngraced commented on mm2src/mm2_main/tests/docker_tests_main.rs:
```diff
@@ -55,6 +55,8 @@ pub fn docker_tests_runner(tests: &[&TestDescAndFn]) {
         const IMAGES: &[&str] = &[
             UTXO_ASSET_DOCKER_IMAGE_WITH_TAG,
             QTUM_REGTEST_DOCKER_IMAGE_WITH_TAG,
+            ZOMBIE_ASSET_DOCKER_IMAGE,
+            ZOMBIE_ASSET_DOCKER_IMAGE_WITH_TAG,
```
delete one of them

borngraced commented on mm2src/coins/z_coin.rs:
```diff
@@ -284,10 +282,14 @@ impl ZCoin {
-            Ok(SyncStatus::Finished { block_number: _, .. })
-        )
+        let mut watcher = self.z_fields.sync_state_connector.lock().await;
+        while let Some(sync) = watcher.sync_watcher.next().await {
```
the formal leaves a deadlock when used in a loop..

shamardy replied:
I will allow it since it's only used in tests, please mark it with the right attributes.

shamardy replied:
> please mark it with the right attributes

This is not fixed yet

shamardy reviewed (dismissed):
LGTM! Please fix any unresolved comments from the previous review to merge this.
